sdk/keyvault/keyvault-keys/test/tracing.test.ts

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { KeyClient, JsonWebKey } from "../src";
+import * as assert from "assert";
+import { TestTracer, setTracer, OperationTracingOptions } from "@azure/core-tracing";
+import { Context as MochaTestContext, Test } from "mocha";
+import { env, Recorder, isRecordMode, isPlaybackMode, delay } from "@azure/test-utils-recorder";
+import { authenticate } from "./utils/testAuthentication";
+import TestClient from "./utils/testClient";
+import { testPollerProperties } from "./utils/recorderUtils";
+import { isNode } from "@azure/core-http";
+
+// This test file confirms that all of our methods that create tracing spans
+// are effectively doing so.
+
+/**
+ * A catch-all type for the SpanGraph root and its nodes.
+ */
+interface TestRecursiveGraph {
+  name?: string;
+  roots?: TestRecursiveGraph[];
+  children?: TestRecursiveGraph[];
+}
+
+/**
+ * Gets the names of all of the nodes of a SpanGraph from @azure/core-tracing
+ */
+function getNames(graph: TestRecursiveGraph): string[] {
+  const { name, roots, children = roots } = graph;
+  if (!children || !children!.length) {
+    return [];
+  }
+  const childrenNames = children.map(getNames).flat();
+  return [name!, ...childrenNames].filter((x) => x);
+}
+
+export type TracingTestFunc = (
+  this: MochaTestContext,
+  tracingOptions: OperationTracingOptions
+) => Promise<void>;
+
+interface TestTracingNodeNames {
+  (nodeNames: string[], test: TracingTestFunc): Test;
+  skip: (nodeNames: string[], test: TracingTestFunc) => Test;
+  only: (nodeNames: string[], test: TracingTestFunc) => Test;
+}
+
+function generateTracingTest(nodeNames: string[], test: TracingTestFunc): Mocha.Func {
```

Comment:
I can see why you referenced decorators, since this sort of inverts and wraps a test. This approach does feel a little more complicated to understand at a glance than a direct "perform operation and assert on results" like other tests.

I feel like another approach would be to pull the before/after parts out into helper methods to minimize the amount of copy/paste, e.g.

```ts
const { rootSpan, tracer } = startTracing();
// do test, pass rootSpan
finishTracing(tracer, rootSpan,  expectedSpans);
```

This might be a little easier to follow/understand?

Reply:
I am not usually strongly opinionated, but decorators would really help us develop more specialized and composable test wrappers around the sample code we're trying to provide inside of the tests.

Though I think the approach you're describing makes complete sense to me. I'll make sure to use it when I go back to this.

Reply:
I'll definitely switch to your approach! You can consider this resolved.